Introduce the `@PyBracketAccess` annotation for using bracket syntax in facades

jlareck opened the pull request.

anatoliykmetyuk commented on coreMacros/src/main/scala/me/shadaj/scalapy/py/Facades.scala:
```diff
@@ -25,18 +25,31 @@ object FacadeImpl {
+      paramss.headOption match {
+        case Some(params) =>
+          val paramExprs = params.map(_.name)
+          paramExprs
```
Looks like this statement does nothing

jlareck commented:
I cleaned up the code and added more options in native_impl method, but I don’t understand in what case we can have None here: https://github.com/shadaj/scalapy/blob/90aeb630738c45ac9419ca6b2f0c30ff1d983371/coreMacros/src/main/scala/me/shadaj/scalapy/py/Facades.scala#L46
Because when I tried to call methods without parameters I got case 0 : https://github.com/shadaj/scalapy/blob/90aeb630738c45ac9419ca6b2f0c30ff1d983371/coreMacros/src/main/scala/me/shadaj/scalapy/py/Facades.scala#L41
Also as we can not use whitebox.context error  because of returned type, I decided to throw ParseException. Maybe it is not the best option but at the moment I don’t know what exception we should throw here in these situations (like if we have 0 parameters or more than 2)

shadaj commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 ## vNEXT
-
+### Highlights :tada:
++ Add support for bracket syntax to facades. The annotation @PyBracketAccess can be used on methods to mark them as representing bracket access on an object. The target method must have one (to read the value) or two parameters (to update the value).
```
Need backticks around `@PyBracketAccess` to format correctly. Also, please refer to the PR at the end of the line like in earlier changelog items.

shadaj commented on core/shared/src/test/scala/me/shadaj/scalapy/py/SpecialSyntaxTest.scala:
```diff
@@ -4,6 +4,15 @@ import me.shadaj.scalapy.interpreter
+  def update(index: Int, newValue: Int): Unit = native
+}
+
+// special syntax test for defining pybrackets
```
Don't need this comment, since we're testing other syntax too.

shadaj commented on core/shared/src/test/scala/me/shadaj/scalapy/py/SpecialSyntaxTest.scala:
```diff
@@ -23,6 +32,15 @@ class SpecialSyntaxTest extends AnyFunSuite {
   }
 
+  test("Can access and update the list elements using brackets") {
+    local{
```
Nit, need a space between `local` and `{`, and remove the extra space between `val` and `myList`.

shadaj commented on coreMacros/src/main/scala/me/shadaj/scalapy/py/Facades.scala:
```diff
@@ -3,8 +3,10 @@ package me.shadaj.scalapy.py
 import scala.reflect.macros.whitebox
 import scala.language.experimental.macros
 
+import scala.reflect.macros.ParseException
```
Unnecessary import?

shadaj commented on coreMacros/src/main/scala/me/shadaj/scalapy/py/Facades.scala:
```diff
@@ -30,13 +32,36 @@ object FacadeImpl {
+          null
+      }
+    }
+    else {
```
`else` should be on the previous line.

shadaj commented on coreMacros/src/main/scala/me/shadaj/scalapy/py/Facades.scala:
```diff
@@ -51,7 +76,7 @@ object FacadeImpl {
     val returnType = method.returnType
     val paramss = method.paramLists
-
+    
```
Nit: extra indent

shadaj commented on docs/static-types.md:
````diff
@@ -31,6 +31,34 @@ If we try to call this method with the wrong parameter type, we get the expected
 string.count(123)
 ```
+### Scala methods representing bracket access
+The annotation @PyBracketAccess can be used on methods to mark them as representing bracket access on an object. The target method must have one (to read the value) or two parameters (to update the value). So, for example, we have a static facade for a list of integers:
````
Need backticks around `@PyBracketAccess`

shadaj replied:
Also, can have a better explanation than `representing bracket access on an object`. Perhaps `representing indexing into the Python object using brackets in Python syntax`?

shadaj replied:
Also, to make the docs more active, we can reword `So, for example, we have` to `For example, we can create`.

shadaj commented on docs/static-types.md:
````diff
@@ -31,6 +31,34 @@ If we try to call this method with the wrong parameter type, we get the expected
+  def update(index: Int, newValue: Int): Unit = py.native
+}
+```
+Then let's create a python list:
````
Nit: capitalize Python

shadaj commented on docs/static-types.md:
````diff
@@ -31,6 +31,34 @@ If we try to call this method with the wrong parameter type, we get the expected
+```scala mdoc
+import py.PyQuote
+
+val  myList = py"[1, 2, 3]".as[IntList]
````
Nit: extra space between `val` and `myList`

shadaj commented on docs/static-types.md:
````diff
@@ -31,6 +31,34 @@ If we try to call this method with the wrong parameter type, we get the expected
+```
+And now we can just use brackets to access elements by indexes. For example, we want to get element by index 0: 
+```scala mdoc
+myList(0) // in python it will call `myList[0]` and return 1
````
Nit: capitalize Python

shadaj commented on docs/static-types.md:
````diff
@@ -31,6 +31,34 @@ If we try to call this method with the wrong parameter type, we get the expected
+
+val  myList = py"[1, 2, 3]".as[IntList]
+```
+And now we can just use brackets to access elements by indexes. For example, we want to get element by index 0: 
````
Nit: `element at index 0` instead of `element by index 0`

shadaj reviewed (changes requested):
First pass, mostly just nits.

shadaj reviewed (approved):
LGTM, great work @jlareck!
